rfd/0xxx-tls-routing-behind-lb.md

```diff
@@ -0,0 +1,319 @@
+ │ │                   └───┬─────┘        └┬────────┘               │      │
+ │ │                       │               │                        │      │
+ │ │Negotiated             │No ALPN        │remote error: tls:      │other │
+ │ │teleport-reversetunnel │Negotiated     │no application protocol │errors│
+ │ │                       │               │                        │      │
+ │ │                   ┌───▼────┐          │                  ┌─────▼──┐   │
+ │ │                   │Upgrade ◄──────────┘                  │  Not   │   │
+ └─┼───────────────────►Required│                             │Required◄───┘
+   │                   └────────┘                             └────▲───┘
+   │                                                               │
+   └───────────────────────────────────────────────────────────────┘
+```
+Connection upgrade for TLS Routing should be required when the client and the server fail to neogiate the ALPN protocol,
+hinting there is a load balancer in the middle that terminates TLS.
+
+However, this test may not be bullet proof. In particular, the test explicitly looks for `tls: no application protocol`
+from the remote when handshakes fails, but it is possible that a load balancer implementation decides to use a different
+text.
+
+In such cases, setting the mode to `on` can enforce the clients to always do the connection upgrade. Also, setting the
+mode `on` or `off` improves performance by skipping the auto test. An environment variable (e.g.
+`TELEPORT_TLS_ROUTING_UPGRADE_MODE=off`) can be set to overwrite the mode per client, just in case.
+
+### Supporting all TLS Routing protocols
+
+In general, in ALL existing places we are dialing the TLS Routing connection, common logic will be added to the
+`ContextDialer` to automatically upgrade the connection when necessary.
+
+However, not all connections are initiated by `tsh`, `teleport`, or a client usingthe API lib. In such cases, a [local
+Teleport
+proxy](https://github.com/gravitational/teleport/blob/master/rfd/0039-sni-alpn-teleport-proxy-routing.md#local-teleport-proxy)
+is required to dial the TLS Routing connection with the upgrade.
+
+Local proxy has already been implemented for Database Access and Application Access. Here are some details how it can be
+implemented for Kubernetes Access.
+
+#### Local proxy for Kubernetes Access
+
+`tsh` will serve a local HTTPS proxy server with a localhost self-signed CA. An ephemeral (per `tsh` command) KUBECONFIG
+will be provided to Kubernetes applications to connect to the local server. This KUBECONFIG points to the local proxy
+using the cluster `server` field, and the config will include static user credentials generated with the local
+self-signed CA.
+
+```
+ ┌────┐                 ┌─────────┐             ┌────────┐
+ │kube│                 │tsh local│             │Teleport│
+ │app │                 │  proxy  │             │ Proxy  │
+ └┬───┘                 └─┬───────┘             └──────┬─┘
```

Comment:
Might make sense to include ALB in this diagram as well? The local proxy will handle connection upgrade?